### PR TITLE
recombine shards method

### PR DESCRIPTION
**Pull request by ameba23**

### CODE REVIEW

**Comment by ameba23** [comment]
still a lot to be done here, but i wanted to start this as i think trying to writing a test for it will hopefully enlighten me to what is going wrong with the invite-reply method

**Comment by ameba23** [comment]
i just cherry-picked the two commits of the rebased branch, as this seemed simplest to me.

what this means is we need to merge 'invites' branch to master (PR awaiting review) then merge this one and it should all work exactly as in the 'recombine-shards-rebase' branch.  

im gonna paste here a comment i put in that branch so that it doesn't get missed:

> i found the problem with verifying

> ssb-keys box is not deterministic. it uses a random salt:
> box('foo',[someid]) != box('foo',[someid])

> i dont see a way around this right now so im gonna take this 'maliciously changed shard' validation out for now, so we have something working until we find a way around the problem.

**Comment by ameba23** [comment]
Just to make it totally clear, this branch currently wont pass tests as the stuff from 'invites' is missing.  if you want to see the tests passing look at 'recombine-rebase', but this is the branch that wants to be merged to master

File: recombine/async/recombine.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    pull(
+      server.query.read(findAssociatedMessages('dark-crystal/ritual')),
+      pull.collect((err, rituals) => {
+        // TODO: verify rituals.length === 1
+        var quorum = getContent(rituals[0]).quorum
+
+        // Get the encrypted shards to use in verification
+        // this is no longer needed but ill leave it here for now
+        pull(
+          server.query.read(findAssociatedMessages('dark-crystal/shard')),
```
[the remark this answers is not in the document]
Comment (ameba23) [reply recombine/async/recombine.js]: thanks.  that usefull.  gonna ditch this bit for now anyway

File: reply/async/reply.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        value: {
+          author: inviteAuthor,
+          content: {
+            root: rootId
+          }
+        }
+      } = invite
+
+      pull(
+        pullShardsByRoot(rootId, { limit: 0 }),
```
Comment (mixmix) [inline reply/async/reply.js]: limit 0 ?
don't know how this behaves, I would have thought it should be `limit: 1`

Comment (mixmix) [reply reply/async/reply.js]: :white_check_mark: fixed

File: request/async/request.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+module.exports = function (server) {
+  const invites = ScuttleInvite(server)
+  const pullShardsByRoot = PullShardsByRoot(server)
+
+  return function request (rootId, callback) {
+    if (!isMsgId(rootId)) return callback(new Error('Invalid root'))
+
+    pull(
+      pullShardsByRoot(rootId, { limit: 0 }),
```
Comment (mixmix) [inline request/async/request.js]: :white_check_mark: removed `limit: 0` , added default empty value to all pull functions (so you don't have to pass opts)

File: shard/pull/byRoot.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        value: {
+          content: {
+            type: 'dark-crystal/shard',
+            root: rootId
+          }
+        }
+      }
+    }]
+
+    const _opts = Object.assign({}, { query }, opts, { limit: 100 })
```
Comment (mixmix) [inline shard/pull/byRoot.js]: as is, this will force the limit to be 100 (as the right-most gets the last say)

:white_check_mark: fixed

**Pull request merged**